src/kernels/jupyter/connection/jupyterConnection.ts

```diff
@@ -144,4 +185,197 @@ export class JupyterConnection {
+                cookieString = (pwSettings.requestHeaders as any).Cookie || '';
+
+                // Password may have overwritten the base url and token as well
+                if (pwSettings.remappedBaseUrl) {
+                    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+                    (serverSettings as any).baseUrl = pwSettings.remappedBaseUrl;
+                    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+                    (serverSettings as any).wsUrl = pwSettings.remappedBaseUrl.replace('http', 'ws');
+                }
+                if (pwSettings.remappedToken) {
+                    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+                    (serverSettings as any).token = pwSettings.remappedToken;
+                }
+            } else if (pwSettings) {
+                serverSettings = { ...serverSettings, token: '' };
+            } else {
+                throw new JupyterInvalidPasswordError();
+            }
+        } else {
+            serverSettings = { ...serverSettings, token: connInfo.token, appendToken: true };
+        }
+
+        const allowUnauthorized = this.configService.getSettings(undefined).allowUnauthorizedRemoteConnection;
+        // If this is an https connection and we want to allow unauthorized connections set that option on our agent
+        // we don't need to save the agent as the previous behaviour is just to create a temporary default agent when not specified
+        if (connInfo.baseUrl.startsWith('https') && allowUnauthorized && this.requestAgentCreator) {
+            const requestAgent = this.requestAgentCreator.createHttpRequestAgent();
+            requestInit = { ...requestInit, agent: requestAgent };
+        }
+
+        // This replaces the WebSocket constructor in jupyter lab services with our own implementation
+        // See _createSocket here:
+        // https://github.com/jupyterlab/jupyterlab/blob/cfc8ebda95e882b4ed2eefd54863bb8cdb0ab763/packages/services/src/kernel/default.ts
+        serverSettings = {
+            ...serverSettings,
+            init: requestInit,
+            WebSocket: this.requestCreator.getWebsocketCtor(
+                cookieString,
+                allowUnauthorized,
+                connInfo.getAuthHeader,
+                connInfo.getWebsocketProtocols?.bind(connInfo)
+                // eslint-disable-next-line @typescript-eslint/no-explicit-any
+            ) as any,
+            fetch: this.requestCreator.getFetchMethod(),
+            Request: this.requestCreator.getRequestCtor(cookieString, allowUnauthorized, connInfo.getAuthHeader),
+            Headers: this.requestCreator.getHeadersCtor()
+        };
+
+        return this.jupyterlab.ServerConnection.makeSettings(serverSettings);
+    }
+
+    // If connecting on HTTP without a token prompt the user that this connection may not be secure
+    private async insecureServerWarningPrompt(): Promise<boolean> {
```

Comment:
Moved from JupyterSessionManager
(unfortunately this introduces a cyclic dependency, but the plan is to fix this)